File: backend_etl/schemas/project.py
from datetime import date as Date, datetime as DateTime
from decimal import Decimal
from typing import Literal, Optional
from urllib.parse import urlparse

from pydantic import BaseModel, Field, JsonValue, field_validator, model_validator


class ProjectCreateRequest(BaseModel):
    nom: str
    dateCreation: Optional[Date] = None
    langueChant: Optional[str] = "Français"
    statutJuridiqueId: Optional[int] = Field(default=None, gt=0)
    statutJuridique: Optional[str] = "Aucun"
    bioCourte: Optional[str] = None


class ProjectCreateResponse(BaseModel):
    id: int


class ProjectUpdateRequest(BaseModel):
    nom: str
    dateCreation: Date
    statutJuridiqueId: Optional[int] = Field(default=None, gt=0)
    statutJuridique: str
    suisaInscrit: bool
    hasLocal: bool
    hasFicheTechnique: bool
    hasMerch: bool
    personalPageUrl: Optional[str] = Field(default=None, max_length=2048)
    personalPageSiteName: Optional[str] = Field(default=None, max_length=100)
    objectifsCourtTerme: Optional[str] = Field(default=None, max_length=5000)
    objectifsMoyenTerme: Optional[str] = Field(default=None, max_length=5000)

    @field_validator(
        "personalPageUrl",
        "personalPageSiteName",
        "objectifsCourtTerme",
        "objectifsMoyenTerme",
    )
    @classmethod
    def strip_optional_project_text(cls, value: Optional[str]) -> Optional[str]:
        return value.strip() or None if value is not None else None

    @model_validator(mode="after")
    def validate_personal_page(self):
        if (self.personalPageUrl is None) != (self.personalPageSiteName is None):
            raise ValueError("Le nom du site et son URL doivent être renseignés ensemble.")
        if self.personalPageUrl is None:
            return self

        parsed_url = urlparse(self.personalPageUrl)
        if parsed_url.scheme not in ("http", "https") or not parsed_url.hostname:
            raise ValueError("L'URL de la page doit utiliser HTTP ou HTTPS.")
        known_hosts = {
            "Instagram": "instagram.com",
            "linktr.ee": "linktr.ee",
            "lift.bio": "lift.bio",
            "hyperfollow.com": "hyperfollow.com",
            "beacons.ai": "beacons.ai",
        }
        expected_host = known_hosts.get(self.personalPageSiteName)
        hostname = parsed_url.hostname.lower()
        if expected_host and hostname != expected_host and not hostname.endswith(f".{expected_host}"):
            raise ValueError(f"L'URL ne correspond pas au site {self.personalPageSiteName}.")
        return self


class ProjectUpdateResponse(BaseModel):
    id: int
    nom: str
    dateCreation: Date
    statutJuridiqueId: int
    statutJuridique: str
    suisaInscrit: bool
    hasLocal: bool
    hasFicheTechnique: bool
    hasMerch: bool
    personalPageUrl: Optional[str] = None
    personalPageSiteName: Optional[str] = None
    objectifsCourtTerme: Optional[str] = None
    objectifsMoyenTerme: Optional[str] = None


class AccompanimentReferenceResponse(BaseModel):
    id: int
    nom: str
    organisme: Optional[str] = None
    estActif: bool


class ProjectFormationWrite(BaseModel):
    formationId: int = Field(gt=0)
    dateSuivi: Date


class ProjectProgrammeWrite(BaseModel):
    programmeId: int = Field(gt=0)
    anneeParticipation: int = Field(ge=1900, le=2100)
    estLaureat: bool


class ProjectFormationResponse(ProjectFormationWrite):
    nom: str
    organisme: Optional[str] = None


class ProjectProgrammeResponse(ProjectProgrammeWrite):
    nom: str
    organisme: Optional[str] = None


class ProjectAccompanimentUpdateRequest(BaseModel):
    formationsSuivies: list[ProjectFormationWrite]
    programmesPrecedents: list[ProjectProgrammeWrite]

    @model_validator(mode="after")
    def reject_duplicate_accompaniment_entries(self):
        formation_keys = {(item.formationId, item.dateSuivi) for item in self.formationsSuivies}
        if len(formation_keys) != len(self.formationsSuivies):
            raise ValueError("Une même formation ne peut apparaître deux fois à la même date.")
        programme_keys = {
            (item.programmeId, item.anneeParticipation)
            for item in self.programmesPrecedents
        }
        if len(programme_keys) != len(self.programmesPrecedents):
            raise ValueError("Un même programme ne peut apparaître deux fois la même année.")
        return self


class ProjectAccompanimentUpdateResponse(BaseModel):
    formationsSuivies: list[ProjectFormationResponse]
    programmesPrecedents: list[ProjectProgrammeResponse]


class AccompanyingExpertResponse(BaseModel):
    id: int
    nom: str
    prenom: str
    email: str
    dateAffectation: Optional[DateTime] = None


class ProjectExpertAssignmentRequest(BaseModel):
    expertId: Optional[int] = Field(default=None, gt=0)


class ProjectImageUpdateRequest(BaseModel):
    url: str


class ProjectImageUpdateResponse(BaseModel):
    avatarUrl: str


class AssetTypeResponse(BaseModel):
    id: int
    libelle: str
    description: Optional[str] = None


class AssetWriteRequest(BaseModel):
    titre: str = Field(min_length=1, max_length=255)
    date: Optional[Date] = None
    typeId: Optional[int] = Field(default=None, gt=0)
    typeLibelle: Optional[str] = Field(default=None, min_length=1, max_length=100)
    url: str = Field(default="", max_length=500)
    metadata: dict[str, JsonValue] = Field(default_factory=dict)
    featuredRoster: bool = False
    isJourneyEvent: bool = False
    eventSource: Optional[str] = Field(default=None, max_length=150)

    @field_validator("titre", "typeLibelle")
    @classmethod
    def strip_required_text(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return None
        value = value.strip()
        if not value:
            raise ValueError("Ce champ ne peut pas être vide.")
        return value

    @field_validator("eventSource")
    @classmethod
    def strip_optional_text(cls, value: Optional[str]) -> Optional[str]:
        return value.strip() or None if value is not None else None

    @field_validator("url")
    @classmethod
    def validate_url(cls, value: str) -> str:
        value = value.strip()
        parsed_url = urlparse(value)
        if value and (parsed_url.scheme not in ("http", "https") or not parsed_url.netloc):
            raise ValueError("L'URL doit utiliser HTTP ou HTTPS.")
        return value

    @model_validator(mode="after")
    def validate_asset(self):
        if (self.typeId is None) == (self.typeLibelle is None):
            raise ValueError("Indiquez soit typeId, soit typeLibelle.")
        if self.isJourneyEvent and self.date is None:
            raise ValueError("La date est obligatoire pour un événement de parcours.")
        if not self.isJourneyEvent:
            self.eventSource = None
        return self


class HighlightResponse(BaseModel):
    id: int
    type: str
    titre: str
    date: Date
    source: Optional[str] = None
    url: Optional[str] = None


class AssetResponse(BaseModel):
    id: int
    titre: str
    date: Optional[Date] = None
    typeId: int
    type: str
    metadata: dict[str, JsonValue]
    url: str
    featuredRoster: bool
    isJourneyEvent: bool
    eventId: Optional[int] = None
    eventSource: Optional[str] = None


class AssetMutationResponse(BaseModel):
    asset: AssetResponse
    highlight: Optional[HighlightResponse] = None


class AssetDeleteResponse(BaseModel):
    assetId: int
    eventId: Optional[int] = None


class TrackOriginalityUpdateRequest(BaseModel):
    estOriginal: bool


class TrackOriginalityUpdateResponse(BaseModel):
    id: int
    estOriginal: bool


class TrackDeleteResponse(BaseModel):
    trackId: int


class EvaluationUpdateRequest(BaseModel):
    campaignId: int = Field(gt=0)
    scores: dict[str, int]
    remarques: str = Field(max_length=10000)
    statut: Literal["en_attente", "selectionne", "refuse"]


class EvaluationUpdateResponse(BaseModel):
    id: int
    campaignId: int
    projectId: int
    scores: dict[str, float]
    remarques: str
    statut: Literal["en_attente", "selectionne", "refuse"]
    decisionEmailSentAt: Optional[DateTime] = None
    gridScore: Optional[float] = None
    gridCoverageCount: int
    gridCriteriaCount: int
    juryAverage: Optional[float] = None
    globalScore: Optional[float] = None
    juryCoverageCount: int
    juryCount: int
    juryNotes: list[dict]


class DecisionEmailSentResponse(BaseModel):
    decisionEmailSentAt: DateTime


class MemberAddRequest(BaseModel):
    nom: str
    prenom: str
    dateNaissance: Date
    genre: str
    npa: str
    ville: str
    email: str
    telephone: Optional[str] = None
    role: str


class MemberResponse(BaseModel):
    id: int
    nom: str
    prenom: str
    genre: str
    dateNaissance: Date
    email: str
    telephone: Optional[str] = None
    role: str
    hasAccount: bool


class VenueCreateRequest(BaseModel):
    nom: str = Field(min_length=1, max_length=150)
    adresse: Optional[str] = Field(default=None, max_length=255)
    ville: str = Field(min_length=1, max_length=100)
    npa: Optional[str] = Field(default=None, max_length=20)
    pays: str = Field(min_length=1, max_length=100)
    jauge: int = Field(ge=0)
    estFestival: bool = False


class VenueResponse(BaseModel):
    id: int
    nom: str
    adresse: Optional[str] = None
    ville: str
    npa: Optional[str] = None
    pays: str
    jauge: int
    estFestival: bool


class ConcertAddRequest(BaseModel):
    date: str
    type: Literal["Local (NE)", "Hors-Canton", "Export"]
    typeEvenement: Literal["Concert", "Résidence", "Showcase"]
    cachetBrut: Optional[Decimal] = Field(default=None, ge=0)
    venueId: Optional[int] = Field(default=None, gt=0)
    newVenue: Optional[VenueCreateRequest] = None


class ConcertResponse(BaseModel):
    id: int
    annee: int
    date: Date
    venueId: int
    lieu: str
    ville: str
    pays: str
    jauge: int
    type: str
    typeEvenement: str
    cachetBrut: Optional[Decimal] = None


class MemberInviteResponse(BaseModel):
    invitationUrl: str

File: frontend/src/types/project.ts
import type { GrantApplication } from '@/types/grantDeadline'

export type JsonValue = string | number | boolean | null | JsonValue[] | { [key: string]: JsonValue }
export type AssetMetadata = { [key: string]: JsonValue }

export interface Member {
  id: number
  nom: string
  prenom: string
  genre: 'Homme' | 'Femme' | 'Autre'
  dateNaissance: string
  role: string
  email: string
  telephone: string
  hasAccount: boolean
}

export interface Track {
  id: number
  titre: string
  duree: string
  dateCreation: string
  isrc: string
  hasAudio: boolean
  estOriginal: boolean
  streamsCount?: string
}

export interface ProjectAsset {
  id: number
  titre: string
  date: string | null
  typeId: number
  type: string
  metadata: AssetMetadata
  url: string
  featuredRoster: boolean
  isJourneyEvent: boolean
  eventId: number | null
  eventSource: string | null
}

export interface AssetType {
  id: number
  libelle: string
  description: string | null
}

export interface AssetWrite {
  titre: string
  date: string | null
  typeId?: number
  typeLibelle?: string
  url: string
  metadata: AssetMetadata
  featuredRoster: boolean
  isJourneyEvent: boolean
  eventSource?: string | null
}

export interface AssetMutationResponse {
  asset: ProjectAsset
  highlight: Highlight | null
}

export interface AssetDeleteResponse {
  assetId: number
  eventId: number | null
}

export interface Venue {
  id: number
  nom: string
  adresse?: string
  ville: string
  npa?: string
  pays: string
  jauge: number
  estFestival: boolean
}

export interface VenueCreate {
  nom: string
  adresse?: string
  ville: string
  npa?: string
  pays: string
  jauge: number
  estFestival: boolean
}

export interface ConcertCreate {
  date: string
  type: 'Local (NE)' | 'Hors-Canton' | 'Export'
  typeEvenement: 'Concert' | 'Résidence' | 'Showcase'
  cachetBrut?: number
  venueId?: number
  newVenue?: VenueCreate
}

export interface Concert {
  id: number
  annee: number
  date: string
  venueId?: number
  lieu: string
  ville: string
  pays: string
  jauge: number
  type: 'Local (NE)' | 'Hors-Canton' | 'Export'
  typeEvenement?: 'Concert' | 'Résidence' | 'Showcase'
  cachetBrut?: number | null
}

export interface Highlight {
  id: number
  type: string
  titre: string
  date: string
  source: string | null
  url: string | null
}

export interface AudienceMetric {
  platform: string
  type: string
  value: number
  date: string
}

export type EvaluationScores = Record<string, number>
export type EvaluationCriterionType = 'objectif' | 'subjectif'
export type EvaluationMode = 'automatique' | 'manuel'

export interface EvaluationCriterionWrite {
  nom: string
  type: EvaluationCriterionType
  noteMaximale: number
  poids: number
  ordre: number
  modeEvaluation: EvaluationMode
  regleObjective: string | null
}

export interface EvaluationCriterion extends EvaluationCriterionWrite {
  id: number
  code: string
  estActif: boolean
  estUtilise: boolean
}

export interface CampaignJuror {
  id: number
  prenom: string
  nom: string
  email: string
}

export type EvaluationCampaignStatus = 'brouillon' | 'ouverte' | 'cloturee'

export interface EvaluationCampaignWrite {
  nom: string
  statut: EvaluationCampaignStatus
  dateDebut: string | null
  dateFin: string | null
}

export interface EvaluationCampaign extends EvaluationCampaignWrite {
  id: number
  dateCreation: string
  projectIds: number[]
  jurorIds: number[]
}

export interface JuryNote {
  expertId: number
  expertNom: string
  note: number | null
  estNotePersonnelle: boolean
}

export interface CampaignEvaluation {
  campaignId: number
  projectId: number
  bilanId: number | null
  scores: EvaluationScores
  remarques: string
  statut: 'en_attente' | 'selectionne' | 'refuse'
  decisionEmailSentAt: string | null
  gridScore: number | null
  gridCoverageCount: number
  gridCriteriaCount: number
  juryAverage: number | null
  globalScore: number | null
  juryCoverageCount: number
  juryCount: number
  juryNotes: JuryNote[]
}

export interface ObjectiveRule {
  code: string
  libelle: string
}

export interface AccompanimentReference {
  id: number
  nom: string
  organisme: string | null
  estActif: boolean
}

export interface ProjectFormation {
  formationId: number
  nom: string
  organisme: string | null
  dateSuivi: string
}

export interface ProjectProgramme {
  programmeId: number
  nom: string
  organisme: string | null
  anneeParticipation: number
  estLaureat: boolean
}

export interface AccompanyingExpert {
  id: number
  nom: string
  prenom: string
  email: string
  dateAffectation?: string | null
}

export interface MusicalProject {
  id: number
  nom: string
  dateCreation: string
  commune?: string
  genreMusical: string
  styles: ProjectStyle[]
  statutJuridiqueId: number
  statutJuridique: string
  langueChant: string
  bioCourte: string
  personalPageUrl?: string | null
  personalPageSiteName?: string | null
  objectifsCourtTerme?: string | null
  objectifsMoyenTerme?: string | null
  pitchAccroche?: string
  bioComplete?: string
  avatarUrl: string
  readinessScore: number
  // Métadonnées & Accompagnement Embrayage
  suisaInscrit: boolean
  hasLocal: boolean
  hasDemandeSubvention: boolean
  hasFicheTechnique: boolean
  hasMerch: boolean
  hasStudioContact?: boolean
  mixMasterContact: string
  formationsSuivies: ProjectFormation[]
  programmesPrecedents: ProjectProgramme[]
  grantApplications: GrantApplication[]
  accompanyingExpert?: AccompanyingExpert | null
  residencesCount: number
  instaFollowers: string
  audienceMetrics: AudienceMetric[]
  // Relations
  members: Member[]
  tracks: Track[]
  assets: ProjectAsset[]
  concerts: Concert[]
  highlights?: Highlight[]
  // Évaluation interne
  scores?: EvaluationScores
  remarquesEvaluation?: string
  statutSelection?: 'en_attente' | 'selectionne' | 'refuse'
  decisionEmailSentAt?: string | null
}

export interface UserProfile {
  id: number
  nom: string
  prenom: string
  email: string
  telephone: string
  genre: string
  dateNaissance: string
  avatarUrl: string
  // Fiche personnelle (table "Personne")
  adresse?: string
  npa?: string
  ville?: string
  canton?: string
  noIpi?: string
}

export interface AccountIdentity {
  id: number
  email: string
  displayName: string
  role: string
  personId: number | null
  canManageEvaluations: boolean
}

export interface ConnectorStatus {
  connected: boolean
  profileUrl?: string | null
  externalId?: string | null
  metrics: ConnectorMetric[]
  metricValue?: number | null
  metricType?: string | null
  lastSync?: string | null
}

export interface ConnectorMetric {
  type: string
  value: number
  date: string
}

export interface SyncConnectors {
  spotify: ConnectorStatus
  mx3: ConnectorStatus
}

export interface Mx3Band {
  id: number
  name: string
  city?: string | null
  profileUrl?: string | null
  imageUrl?: string | null
}

export interface Mx3Gig {
  name: string
  date: string
  bandName?: string | null
  stageName?: string | null
  location?: string | null
  locationUrl?: string | null
  ticketUrl?: string | null
}

// Taxonomie des styles musicaux (arbre à 3 niveaux : racine FCMA -> macro-genre -> sous-genre)
export interface StyleMusical {
  id: number
  nom: string
  idParent: number | null
}

// Style attribué à un projet (feuille de l'arbre StyleMusical), avec indicateur "principal"
export interface ProjectStyle {
  id: number
  nom: string
  principal: boolean
}

// Catalogue Spotify récupéré à la volée (non persisté), pour sélection avant import en base
export interface SpotifyCatalogTrack {
  id: string
  name: string
  track_number: number
  duration_ms: number
  release_date: string
}

export interface SpotifyCatalogAlbum {
  id: string
  name: string
  release_date: string
  album_type: string
  tracks: SpotifyCatalogTrack[]
}

export interface SpotifyCatalog {
  artist_id: string
  artist_name: string
  followers: number
  popularity: number
  genres: string[]
  albums: SpotifyCatalogAlbum[]
}

export interface LegalStatus {
  id: number
  libelle: string
  estActif: boolean
}